numpy/core/src/umath/scalarmath.c.src

```diff
@@ -569,6 +571,92 @@ static void
 
 /*** END OF BASIC CODE **/
 
+/**
+ * Find the descriptor for a builtin or Python numerical scalar.
+ * This function should only be used for scalar math fast paths.
+ *
+ * @param obj The object for which to find the descriptor
+ * @param descr The descriptor that was found.
+ * @returns 1 if the other object is a Python type and thus must not
+ *          handle the operation, otherwise 0. -1 _without_ an error set
+ *          if no descriptor was found.
+ */
+static int
+descr_from_basic_scalar(PyObject *obj, PyArray_Descr **descr)
+{
+    /* TODO: We could try giving defined scalars a chance... */
+    int type_num = PyArray_TypeNumFromNumPyScalarType(
+            (PyObject *)Py_TYPE(obj), 0);
+    if (type_num != NPY_NOTYPE) {
+        *descr = PyArray_DescrFromType(type_num);
+        return 0;
+    }
+    else if (PyFloat_CheckExact(obj)) {
+        *descr = PyArray_DescrFromType(NPY_DOUBLE);
+        return 1;
+    }
+    else if (PyLong_CheckExact(obj)) {
+        if (PyLong_AsLong(obj) == -1 && PyErr_Occurred()) {
+            PyErr_Clear();
+            return -1;
+        }
+        *descr = PyArray_DescrFromType(NPY_LONG);
+        return 1;
+    }
+    else if (PyComplex_CheckExact(obj)) {
+        *descr = PyArray_DescrFromType(NPY_CDOUBLE);
+        return 1;
+    }
+    return -1;
+}
+
+
+/**
+ * This function attempts to compare NumPy or Python scalars
+ * The operation is done by getting value of both scalars
+ * and calling richcompare on them.
+ *
+ * @param self The first object which will be converted to scalar item
+ * @param other The second object to compare to
+ * @param other The value of comparison operator
+ * @returns Py_False or Py_True if richcompare is successfull
+ *          , otherwise NULL.
+ */
+static PyObject*
+do_richcompare_on_scalars(PyObject *self, PyObject *other, int cmp_op) {
+    PyObject *item_self, *ret=NULL;
+    PyArray_Descr *self_descr=NULL, *other_descr=NULL;
+    void *data_self;
+    int pyscalar_other, pyscalar_self;
+    int is_complex_operands, is_equality_operator;
+
+    pyscalar_self = descr_from_basic_scalar(self, &self_descr);
+    pyscalar_other = descr_from_basic_scalar(other, &other_descr);
+
+    if (pyscalar_self >= 0 && pyscalar_other >= 0) {
+        /*
+         * If either of the operands are complex AND the operator is not equality,
```

Comment:
```suggestion
         * If both operands are complex AND the operator is not equality,
```
meant this, will fix

Reply:
Ok this comment is actually right the logic is incorrect :) , will fix it.

Reply:
Thanks! Just make sure we have a test, that would be great ;)  (I admit it can be pretty annoying in this type of context where the tests exist a bit, but not thorough.)

Reply:
Luckily I found this via a test(https://github.com/numpy/numpy/pull/17970/checks?check_run_id=1767648779#step:4:5730), but yeah, I'll see how to make it a bit more thorough